File: easy/array/non_decreasing_array/non_decreasing_array.py
class Solution(object):
    def checkPossibility(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        place_of_violation = -1

        for i in range(0, len(nums)-1):
            if nums[i] > nums[i+1]:
                if place_of_violation != -1:
                    return False
                place_of_violation = i

        if place_of_violation == -1:
            return True

        index_to_change = place_of_violation
        if index_to_change == 0:
            return True
        elif nums[index_to_change-1] <= nums[index_to_change+1]:
            return True

        index_to_change = place_of_violation + 1
        if index_to_change == len(nums)-1:
            return True
        elif nums[index_to_change-1] <= nums[index_to_change+1]:
            return True

        return False

File: easy/array/non_decreasing_array/non_decreasing_array.go
package main

import "fmt"

func checkPossibility(nums []int) bool {
	placeOfViolation := -1

	for i := 0; i < len(nums)-1; i++ {
		if nums[i] > nums[i+1] {
			if placeOfViolation != -1 {
				return false
			}
			placeOfViolation = i
		}
	}

	if placeOfViolation == -1 {
		return true
	}

	indexToChange := placeOfViolation
	if indexToChange == 0 {
		return true
	}

	if nums[indexToChange-1] <= nums[indexToChange+1] {
		return true
	}

	indexToChange = placeOfViolation + 1
	if indexToChange == len(nums)-1 {
		return true
	}

	if (indexToChange < len(nums)-1) && (nums[indexToChange-1] <= nums[indexToChange+1]) {
		return true
	}

	return false
}

func main() {
	fmt.Println("Solution 1 : ", checkPossibility([]int{4, 2, 3}))
	fmt.Println("Solution 2 : ", checkPossibility([]int{4, 2, 1}))
}
